pyquil/quilbase.py

```diff
@@ -413,6 +413,36 @@ def num_args(self) -> int:
         return int(np.log2(len(self.permutation)))
 
 
+class DefGateByPaulis(DefGate):
+    """
+    Records a gate definition as the exponentiation of a PauliSum.
+    """
+
+    # actually, body is of type PauliSum, but circular imports are no good
+    def __init__(self, gate_name: str, parameters: list, arguments: list, body):  # type: ignore
+        if not isinstance(gate_name, str):
+            raise TypeError("Gate name must be a string")
+
+        if gate_name in RESERVED_WORDS:
+            raise ValueError("Cannot use {} for a gate name since it's a reserved word".format(gate_name))
+
+        self.name = gate_name
+        self.parameters = parameters
+        self.arguments = arguments
+        self.body = body
+
+    def out(self) -> str:
+        out = f"DEFGATE {self.name}({', '.join(map(str, self.parameters))}) "  # type: ignore
```

Comment:
I think if it were up to me, I'd prefer `DefGate`'s `parameters` member to be nonoptional. I don't see what signal `None` has over `[]`.

But, since that's not presently a fight I want to pick, I'll run with this patch.

Reply:
> I think if it were up to me, I'd prefer `DefGate`'s `parameters` member to be nonoptional. I don't see what signal `None` has over `[]`....

Agreed. If I had to guess, the original intention was probably just to allow the `__init__` arg to be optional.